### ISSUE TITLE
Evaluate model performance when using chainer.links.BatchNormalization togather with chainer.config.train=False bug report

### ISSUE THREAD

**Issue by sharpstill** [chainer/chainer#4553]
chainer version: 3.4.0
cupy version: 2.4.0

When I am using resnet-101 to classify images. The `resnet-101` model have `L.BatchNormalization` layer.
After I trained over, I load model parameter and use `chainer.using_config('train', False)` to evaluate the model performance.
I am surprised that even I am using train dataset(not validation dataset), the accuracy is lower than When I was in training procedure observation ( it was 99%  in the final iteration of training). It was only `80%` in train split dataset.
I think the `L.BatchNormalization` has bug when you set `chainer.using_config('train', False)` to evaluate the pretrained model.

**Comment by kmaehashi** [comment]
Could you share the minimum code to reproduce the issue?

**Comment by jinjiren** [comment]
I think the problem is caused by that `BatchNorm` uses different statistics for training and testing.

My answer is based on the assumption that you are applying a pre-trained model on a new dataset (including training/validation/test set). Maybe I'm wrong 😅

Specifically, if you use a pre-trained model, then the statistics of the batches in the original dataset (maybe ImageNet) is reused. As a result, during training, the statistics (mean, std) is actually the combination of the previous dataset and your current training split. Then if you evaluate the training split again with `chainer.using_config('train', False)`, the statistics is reset and thus is purely from the training split. These differences may cause the performance degradation, as I have previously encountered.

Anyway, I think it's important to consider **what data are used for computing** the running average of statistics for `BatchNorm`, since this will make a big difference for evaluation even if using the same data.

**Comment by sharpstill** [comment]
@jinjiren Yes I use pretrained imagenet model to load and to train.
After I trained over, How should I evaluate the model?? Should I don't set `chainer.using_config('train', False)`?? 
what is your advice

**Comment by sharpstill** [comment]
@jinjiren Do you mean if I load pretrained imagenet weight to train a new dataset , What is the best practice to set chainer.links.BatchNormalization  when evaluate on validation dataset???

**Comment by jinjiren** [comment]
It's better to use `chainer.using_config('train', False)` in this situation. However, this is not a problem specific to Chainer. Instead, it's a common problem when finetuning models with BatchNormalization layers. For more details, you can refer to [1] Y. Li, N. Wang, J. Shi, X. Hou, and J. Liu, “Adaptive Batch Normalization for practical domain adaptation,” Pattern Recognit., vol. 80, pp. 109–117, 2018.

**Comment by sharpstill** [comment]
@jinjiren I am PhD student, I need to publish paper, Right now, I use `chainer.using_config('train', False)` to evaluate my model, but because I use imagenet to pretrain model, Like you say, it will cause accuracy very very low in test dataset.
What should I do?? Must I retrain the model without load ImageNet parameter?

**Comment by sharpstill** [comment]
Because re-train is time consuming and it will cause I take a week. So it is expensive

**Comment by jinjiren** [comment]
Sorry I made a typo, you should use train=True when evalutation for this task.

**Comment by sharpstill** [comment]
@jinjiren But, please see my stackoverflow page: https://stackoverflow.com/questions/49808312/how-to-use-chainer-links-batchnormalization-when-loading-pretrained-parameter-th
@cocuh written:
```
In testing mode, it uses pre-computed population statistics to normalize
the input variable. The population statistics is approximated if it is
computed by training mode, or accurate if it is correctly computed by
fine-tuning mode.
```
He has different opinion with yours.

**Comment by jinjiren** [comment]
By the way, bad experiment results are not other people’s faults. Please be polite to people who are helping you. It’s your own responsibility to do good research, understand and analyze the experiment results. At least you should read the paper I linked carefully and probably discuss or cite it in your paper. 

Deadline is not the reason to be rude.

**Comment by sharpstill** [comment]
@jinjiren I am very very sorry to offend you , Please forgive me, I am appreciate your generous help!
I am urgent in deadline, If I say something wrong, please forgive me.
Again sorry,  I didn't mean to rude

**Comment by sharpstill** [comment]
I will read carefully your given paper. After that I will talk to you, Will you help me please.

**Comment by sharpstill** [comment]
@unnonouno Can you help me about this issue please, I am stuck it for weeks.

**Comment by sharpstill** [comment]
@jinjiren I have read paper : [1] Y. Li, N. Wang, J. Shi, X. Hou, and J. Liu, “Adaptive Batch Normalization for practical domain adaptation,” Pattern Recognit., vol. 80, pp. 109–117, 2018.
It seems I need to re-implement a parameter-free layer AdaBatch and re-train it.
1. Do you have code of AdaBatch.?
2. Do you mean currently the best practice is to use `chainer.using_config('train', True)` in Chainer?

**Comment by sharpstill** [comment]
@jinjiren Thank you very much!

**Comment by kmaehashi** [comment]
We discussed this topic and will document more on modes of BatchNormalization and when to use each of them.

**Comment by kmaehashi** [comment]
> We discussed this topic and will document more on modes of BatchNormalization and when to use each of them.

I raised another issue for this point #5277.
As for questions, please post on StackOverflow with chainer tag.
